### PR TITLE
feat(evals): B2–B4 — gate metrics, live-run join, corpus growth (closes stack onto pre-0.0.1)

### PR DESCRIPTION
**Pull request by alexhawat**
## Summary

Merges the rest of the eval-benchmark stack (spinning out of #140) onto `pre-0.0.1`. #211 (B1) already merged directly into trunk; #212 (B2), #213 (B3), and #214 (B4) merged into their parent branches instead of trunk, so their code never actually landed on `pre-0.0.1` — this PR closes that gap by merging the stack's tip (`wave/bench-b4-corpus`, which contains all of B1+B2+B3+B4) straight into trunk.

- **B2** (#212): directional gate metrics (`unsafe_approval_rate`/`clean_block_rate`/`inconclusive_rate`) + completed `VersionPins` (N6).
- **B3** (#213): the live finding-location join — `evals/live_run.py`, `mergecraft eval bench`, `make bench-detect`.
- **B4** (#214): corpus grown 10 → 46 bank cases, patch-bearing detection corpus seeded (43 cases) at `evals/bench/mergecraft/`.

All three already went through the full test-creator → implement → wave-verifier cycle (see #212/#213/#214 for verification detail) and were independently merged. This PR is pure consolidation — no new changes beyond what's already reviewed and merged into the stack.

## Test plan

- [x] `make ci` green on `wave/bench-b4-corpus` (verified per #214) modulo the 4 pre-existing unrelated Rich-ANSI test failures already isolated across B1–B4
- [x] `MERGECRAFT_PYTEST_JOBS=0 uv run pytest tests/evals -q` — 156 passed, 1 xfailed

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by alexhawat** [comment]
Reopening from wave/bench-b3-live instead — that's the branch that actually received GitHub's merge commit for #214, not wave/bench-b4-corpus. Content is identical either way, but wave/bench-b3-live is the correct/current ref.

File: src/mergecraft/evals/live_run.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+            asyncio.run(
+                run_offline_diff_review(
+                    cwd=Path.cwd(),
+                    diff_file=case.patch_path,
+                    model=model,
+                    json_path=json_path,
+                )
+            )
+            if not json_path.is_file():
+                return []
```
Comment (sevn-one) [inline src/mergecraft/evals/live_run.py:182]: Returning `[]` when no findings file was written conflates "reviewer found nothing" with "reviewer never ran" — the `OfflineReviewResult` returned by `asyncio.run(run_offline_diff_review(...))` (which carries `success`/`error`) is discarded here. A failed subprocess then scores as a vacuous clean case: recall 1.0 on a clean case, and a false negative on a buggy one. That is exactly the failure mode the structural side's `inconclusive` bucket (via `case.run_succeeded`) was designed to keep out of the published numbers, and the detection side has no equivalent. Consider checking `result.success` and recording a per-case error/skip instead of returning `[]`; at minimum, add a test that drives `run_live_detection` with a stub returning an error/empty payload to pin the intended behavior.

**Review by sevn-one — CHANGES_REQUESTED** [review]
Reviewed the full consolidation. The corpus growth (46 bank / 43 detection cases), the gate-matrix directional aggregation, and the `live_run` join are well documented and the metric math is internally consistent with the new tests. But as committed the feature cannot run at all, so this needs changes before merge.

## Blocking — Python 2 `except` syntax in the module this PR is built around

`src/mergecraft/evals/benchmark.py:258` and `:271`:

```python
except OSError, subprocess.CalledProcessError:
```

This is the Python 2 comma form of `except`. The project requires Python `>=3.14` (`pyproject.toml:10`), under which this is a `SyntaxError: invalid syntax`, raised at compile time on the first import. As a result `mergecraft.evals.benchmark` cannot be imported, which breaks:

- `mergecraft.evals.live_run` (new in this PR — it imports `run_structural_replay` and friends from `benchmark`),
- `mergecraft.cli.eval_cmd` → `replay-bank --gate` and the new `bench` command,
- every `tests/evals` file that imports `benchmark`/`live_run` (collection fails before a single test runs).

`evals/scoring.py:346,351,355` carries the same pattern (`except TypeError, ValueError:`). It is pre-existing rather than added here, but the new `live_run.py` imports `score_findings`/`load_baseline_issues`/`load_reported_findings` from it, so both files must be fixed for this stack to import. The repo's own `src/mergecraft/review_checks.py:12` documents this exact `except A, B:` construct as a syntax error — worth reconciling why nothing in the pipeline actually enforces that.

```python
except (OSError, subprocess.CalledProcessError):
except (TypeError, ValueError):
```

I can't reconcile this with the PR's "make ci green / 156 passed" claims: this tree cannot parse, so no lint/import/collection gate could have passed on it (the only completed-success checks on the head are changelog-bot runs). Please fix these lines, run the suite for real, and re-push before merge.

## Major — a failed live review is scored as "clean" (see inline comment on `live_run.py`)

## Minor — committed result sets are not reproducible as-is

`evals/results/latest.json` and the `structural-replay-*.json` snapshots encode `mergecraft_commit: d4527cf` (the head's parent) and were produced against a codebase state that currently cannot be imported. After the syntax fix, re-run `mergecraft eval bench` and `replay-bank --gate` and commit fresh outputs so the published numbers are reproducible from the committed tree.

---
*via mergecraft* · `nous/deepseek/deepseek-v4-flash` · [run](https://github.com/alexhawat/mergeCraft/actions/runs/31877000637)

**Comment by sevn-one** [comment]
Reviewed PR #215 (B2–B4 eval-benchmark consolidation onto pre-0.0.1). Submitted a CHANGES_REQUESTED review (no prior threads existed; no approval given).

Key findings:
- **Blocking:** `src/mergecraft/evals/benchmark.py:258,271` use Python 2 `except OSError, subprocess.CalledProcessError:` syntax — a `SyntaxError` under the project's required Python >=3.14, so `benchmark` (and transitively the new `live_run`, `eval_cmd`, and all of `tests/evals`) cannot be imported. `evals/scoring.py:346,351,355` has the same pattern and is imported by the new `live_run.py`. Verified via file content (read/grep/git-show/git cat-file) and an independent subagent. This contradicts the "CI green / 156 passed" claims; the only green checks on the head are changelog-bot runs. Fix: `except (OSError, subprocess.CalledProcessError):` / `except (TypeError, ValueError):`.
- **Major (inline on live_run.py:182):** `_default_review_fn` returns `[]` when the diff-review subprocess writes no findings file, scoring a failed review as "clean" (vacuous recall) — the detection side has no inconclusive/skip equivalent to the structural side's `case.run_succeeded` handling.
- **Minor:** committed `evals/results/*.json` encode `mergecraft_commit: d4527cf` (parent of head) from a codebase state that cannot currently import; regenerate after the syntax fix.

Also verified the gate-matrix/directional-rate logic and its tests are internally consistent, and that the eval-bank corpus counts (46 bank / 43 detection) match the committed result sets.

---
*via mergecraft* · `nous/deepseek/deepseek-v4-flash` · [run](https://github.com/alexhawat/mergeCraft/actions/runs/31877000637)

File: src/mergecraft/evals/benchmark.py [hunk 2]
```diff
@@ -145,6 +284,21 @@ def _judge_pins(providers: tuple[str, ...]) -> dict[str, dict[str, Any]]:
...
+    """Pin the reviewing-model identity per provider (N6, D12).
+
+    Mirrors ``_judge_pins``: structural replay never invokes a live provider,
+    but the *configured* pin for each named provider is itself a fact worth
+    recording — a report that later adds a live finding-location run (B3) can
+    then diff its own recorded model against this structural baseline.
+    """
+    pins: dict[str, ReviewingModelPin] = {}
+    for provider in providers:
+        model = pinned_judge_model(provider) or "unknown"
```
Comment (sevn-one) [inline src/mergecraft/evals/benchmark.py:297]: _Data Integrity & Atomicity_ | _Major_ | _Quick win_ | _certain_

These required reviewing-model pins come from the verifier-judge configuration rather than the review models: the default output already serializes OpenAI as `model_id: "unknown"` / `model_pin: "unknown"` and duplicates the Claude alias into both fields. Published results therefore do not carry the reproducible per-provider reviewer identity promised by D12; populate real model IDs/pins and cover the default `run_structural_replay()` output.

File: src/mergecraft/evals/live_run.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    review_fn: ReviewFn,
+    results_dir: Path,
+    slack: int = DEFAULT_LINE_SLACK,
+) -> DetectionMetrics:
+    """Drive every case through ``review_fn``, score it, and fold the results.
+
+    Raw findings are persisted per case under ``results_dir/raw-findings/`` —
+    a published number must be able to show its work (D9).
+    """
+    raw_dir = results_dir / "raw-findings"
```
Comment (sevn-one) [inline src/mergecraft/evals/live_run.py:120]: _Data Integrity & Atomicity_ | _Major_ | _Quick win_ | _certain_

Every run writes the same `results_dir/raw-findings/<case>.json` names while result sets themselves are timestamped. A later run overwrites the evidence behind earlier metrics, so namespace raw findings per run and add coverage showing two consecutive runs preserve both artifacts.

**Review by sevn-one — CHANGES_REQUESTED** [review]
Re-reviewed the current head. The existing unresolved live-review failure-path thread still applies, and the latest commit only repairs two corpus fixtures. The two publication-integrity issues below also need to be addressed before merge.

---
*via mergecraft* · `openai/gpt-5.6-sol` · [run](https://github.com/alexhawat/mergeCraft/actions/runs/31877000637)
